wpilibNewCommands/src/test/native/cpp/frc2/command/CommandScheduleTest.cpp

```diff
@@ -95,6 +101,92 @@ TEST_F(CommandScheduleTest, SchedulerCancel) {
   EXPECT_FALSE(scheduler.IsScheduled(&command));
 }
 
+TEST_F(CommandScheduleTest, CancelNextCommand) {
+  CommandScheduler scheduler = GetScheduler();
+
+  frc2::RunCommand* command1PtrPtr = nullptr;
+  frc2::RunCommand* command2PtrPtr = nullptr;
+  auto counterPtr = std::make_shared<int>(0);
+
+  auto command1 = frc2::RunCommand([counterPtr, &command2PtrPtr, &scheduler] {
+    scheduler.Cancel(command2PtrPtr);
+    (*counterPtr)++;
+  });
+  auto command2 = frc2::RunCommand([counterPtr, &command1PtrPtr, &scheduler] {
+    scheduler.Cancel(command1PtrPtr);
+    (*counterPtr)++;
+  });
+
+  command1PtrPtr = &command1;
+  command2PtrPtr = &command2;
+
+  scheduler.Schedule(&command1);
+  scheduler.Schedule(&command2);
+  scheduler.Run();
+
+  EXPECT_EQ(*counterPtr, 1)
+      << "Second command was run when it shouldn't have been";
+
+  // only one of the commands should be canceled.
+  EXPECT_FALSE(scheduler.IsScheduled(&command1) &&
+               scheduler.IsScheduled(&command2))
+      << "Both commands are running when only one should be";
+  // one of the commands shouldn't be canceled because the other one is canceled
+  // first
+  EXPECT_TRUE(scheduler.IsScheduled(&command1) ||
+              scheduler.IsScheduled(&command2))
+      << "Both commands are canceled when only one should be";
+
+  scheduler.Run();
+  EXPECT_EQ(*counterPtr, 2);
+}
+
+TEST_F(CommandScheduleTest, CommandKnowsWhenItEnded) {
+  CommandScheduler scheduler = GetScheduler();
+
+  frc2::FunctionalCommand* commandPtr = nullptr;
+  auto command = frc2::FunctionalCommand(
+      [] {}, [] {},
+      [&](auto isForced) {
+        EXPECT_FALSE(scheduler.IsScheduled(commandPtr))
+            << "Command shouldn't be scheduled when its end is called";
+      },
+      [] { return true; });
+  commandPtr = &command;
+
+  scheduler.Schedule(commandPtr);
+  scheduler.Run();
+  EXPECT_FALSE(scheduler.IsScheduled(commandPtr))
+      << "Command should be removed from scheduler when its isFinished() "
+         "returns true";
+}
+
+TEST_F(CommandScheduleTest, ScheduleCommandInCommand) {
+  CommandScheduler scheduler = GetScheduler();
+  int counter = 0;
+  frc2::InstantCommand* commandPtr = nullptr;
+
+  auto command = frc2::RunCommand([&] {
+    scheduler.Schedule(commandPtr);
+    EXPECT_EQ(counter, 1)
+        << "Command 2's init was not run immediately after getting scheduled";
+  });
+  auto commandToGetScheduled = frc2::InstantCommand([&counter] { counter++; });
+
+  commandPtr = &commandToGetScheduled;
```

Comment:
```suggestion
  frc2::InstantCommand commandToGetScheduled{[&counter] { counter++; }};

  auto command = frc2::RunCommand([&] {
    scheduler.Schedule(&commandToGetScheduled);
    EXPECT_EQ(counter, 1)
        << "Scheduled cmmand's init was not run immediately after getting scheduled";
  });
```
See Java. You'll also need to update the usages of `commandPtr` in the `IsScheduled` checks.

Reply:
## Added in 93ad7b8